ocaml/xapi/certificates_sync.ml

```diff
@@ -29,16 +29,29 @@ let install ~__context ~host:_ ~type' cert =
     error "certificates_sync.install exception: %s" (Printexc.to_string e) ;
     Error (`Msg ("installation of host certificate failed", []))
 
+type to_update = Certificate | Hashes of {sha256: string; sha1: string}
+
 (** determine if the database is up to date by comparing the fingerprint
   of xapi-ssl.pem with the entry in the database *)
-let is_unchanged ~__context cert_ref cert =
+let to_update ~__context cert_ref cert =
   let ref_hash =
     Db.Certificate.get_fingerprint_sha256 ~__context ~self:cert_ref
   in
-  let cert_hash =
-    X509.Certificate.fingerprint `SHA256 cert |> Certificates.pp_hash
-  in
-  cert_hash = ref_hash
+  let cert_hash = Certificates.pp_fingerprint_of_cert ~hash_type:`SHA256 cert in
+  if ref_hash = "" then
+    (* We must be upgrading from a version predating fingerprint_sha256, so check fingerprint instead *)
+    if cert_hash = Db.Certificate.get_fingerprint ~__context ~self:cert_ref then
+      let sha256 =
+        Certificates.pp_fingerprint_of_cert ~hash_type:`SHA256 cert
+      in
+      let sha1 = Certificates.pp_fingerprint_of_cert ~hash_type:`SHA1 cert in
+      Some (Hashes {sha256; sha1})
```

Comment:
@psafont is this sort of what you had in mind?

Reply:
Yes, this now reads much better on the caller side, thanks